### PR TITLE
Add Typing for the typescript

### PR DESCRIPTION
**Pull request by NaveenDA**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        left:Complex|Compound;
+        type:"complex"
+    }
+    interface Compound{
+        type:"compound";
+        list:TokenizeOutput[]
+    }
+    interface ParserOutput{
+        type:string;
+        combinator:string;
```
Comment (LeaVerou) [inline index.d.ts:29]: `combinator` is also optional

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+    // Type definitions for parsel-js 1.0.1
+    // Project: https://github.com/LeaVerou/parsel
+    // Definitions by: Naveen DA <https://github.com/NaveenDA>
+    // Definitions: https://github.com/LeaVerou/parsel/blob/master/index.d.ts
```
Comment (LeaVerou) [inline index.d.ts:4]: The entire file appears to be indented? Also, would it be possible to use tabs for indentation, like the rest of the project?

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+
+    interface TokenizeOutput{
```
Comment (LeaVerou) [inline index.d.ts:7]: Should this be `TokenizedOutput`? Or even better, just `Tokens`. Verbs should be avoided in naming arguments.

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        combinator:string;
+        right?:Complex|Compound|TokenizeOutput;
+        left:Complex|Compound;
+        type:"complex"
+    }
+    interface Compound{
+        type:"compound";
+        list:TokenizeOutput[]
+    }
+    interface ParserOutput{
```
Comment (LeaVerou) [inline index.d.ts:27]: Wouldn't `AST` be a better name?

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Traverse all tokens of a (sub)tree:
+     */
+    export function walk(node: ParserOutput,callback:(node: ParserOutput)=>{}):void ;
+
+    /**
+     * Calculate specificity (returns an array of 3 numbers):
+     */
+    export function specificity(selector: string, options: SpecificityOptions):number[];
```
Comment (LeaVerou) [inline index.d.ts:62]: Note that `selector` can also be an AST.

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Get list of tokens as a flat array:
+     */
+    export function tokenize(selector: string):TokenizeOutput[] ;
+
+    /**
+     * Traverse all tokens of a (sub)tree:
+     */
+    export function walk(node: ParserOutput,callback:(node: ParserOutput)=>{}):void ;
```
Comment (LeaVerou) [inline index.d.ts:57]: callback also accepts the parent node as a second argument.

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Calculate specificity (returns an array of 3 numbers):
+     */
+    export function specificity(selector: string, options: SpecificityOptions):number[];
+    
+    /**
+     *  To convert the specificity array to a number
+     */
+    export function specificityToNumber(specificity:any[],base?:number):number;
```
Comment (LeaVerou) [inline index.d.ts:67]: shouldn't this be `number[]` instead of `any[]`?

**Review by LeaVerou — CHANGES_REQUESTED** [review]
Thank you! I've left some comments.

File: index.d.ts (hunk 2 above)
Comment (NaveenDA) [reply index.d.ts]: Yeah, it seems like indented. Now fixed

File: index.d.ts (hunk 4 above)
Comment (NaveenDA) [reply index.d.ts]: Yes AST should be the name

**Comment by NaveenDA** [comment]
@LeaVerou thanks for the valuable comments. Now all the issues are fixed.
Please review it.

**Review by LeaVerou — CHANGES_REQUESTED** [review]
Thank you!

I missed this earlier: The AST can be any type of node, not necessarily a complex or compound. It can also just be id, or class, or attribute etc., whose attributes are missing. I would skip the `Compound` type, and just add optional properties on `AST`. And maybe rename to `ASTNode` to make it clearer.

Nit: Code is still using spaces for indentation.

**Comment by NaveenDA** [comment]
So we need to replace `ASTNode` into `ASTNode | string` ?
```js
export function walk(node: ASTNode,callback:(node: ASTNode,parentNode: ASTNode)=>{}):void ;
```

```js
export function walk(node: ASTNode|string,callback:(node: ASTNode|string,parentNode: ASTNode|string)=>{}):void ;

```

**Comment by LeaVerou** [comment]
No, not at all. These functions don't accept a selector, only AST nodes.
What I was saying is, there are other keys the AST Node could contain, it's not necessarily a compound or complex token. E.g. if the input is just `#foo`, the output will be a `{
	"type": "id",
	"content": "#foo",
	"name": "foo",
	"pos": [
		0,
		4
	]
}`, however there is no `name` or `content` keys in `ASTNode`.

**Comment by mattflux** [comment]
> No, not at all. These functions don't accept a selector, only AST nodes.
> What I was saying is, there are other keys the AST Node could contain, it's not necessarily a compound or complex token. E.g. if the input is just `#foo`, the output will be a `{ "type": "id", "content": "#foo", "name": "foo", "pos": [ 0, 4 ] }`, however there is no `name` or `content` keys in `ASTNode`. Same for attribute selectors, class selectors, element selectors.

@LeaVerou I was looking at this and seeing what can be done to introduce these typings. 

Would it suffice to just redefine AST as a union of `Complex | Compound | Tokens` as they are defined here?

**Comment by LeaVerou** [comment]
Yes, I think so!

**Pull request closed without merging**